### PR TITLE
Optimizing partial applications by wrapping them in lambdas if safe

### PR DESCRIPTION
**Pull request by TIHan**
Resolves this: https://github.com/dotnet/fsharp/issues/8240

The strategy here is straight-forward. Simply wrap a partially applied function in a lambda expression; this allows us to take advantage of the optimizations that only work on lambda expressions.

### CODE REVIEW

File: src/fsharp/Optimizer.fs [hunk 1]
```diff
@@ -2968,6 +2970,48 @@ and OptimizeLambdas (vspec: Val option) cenv env topValInfo e ety =
                  HasEffect=false
                  MightMakeCriticalTailcall = false
                  Info= valu }
+
+    // For partial applications, try to wrap it in a lambda and re-optimize as lambdas have better optimization strategies.
+    | Expr.App (Expr.Val (vref, _, _) as vexpr, fty, tinst, args, mApp) when isFunTy g ety ->
+        match vref.ValReprInfo, vspec with
+        | Some valInfo, Some v ->
+        //    let argTys, retTy = stripFunTy g ety
```
Comment (forki) [inline src/fsharp/Optimizer.fs:2978]: commented code

File: src/fsharp/Optimizer.fs [hunk 2]
```diff
@@ -2968,6 +2970,48 @@ and OptimizeLambdas (vspec: Val option) cenv env topValInfo e ety =
...
+                    System.Console.WriteLine
+
+                let test2 a b c d e f g xs =
+                    let action = test a b c d e f g
+                    printfn "world"
+                    xs |> Array.iter action
+            *)
+            // In the example, we should not make a lambda around "test" in "test2" because "test" is fully applied but returns
+            //     a partial application.
+            if valInfo.NumCurriedArgs <> 0 && not v.IsMutable && not vref.IsMutable &&
```
Comment (forki) [inline src/fsharp/Optimizer.fs:2995]: valInfo.NumCurriedArgs is used twice and can be put into a variable

Comment (NinoFloris) [reply src/fsharp/Optimizer.fs:2995]: Most likely CSE will take care of that at runtime, forced locals aren't free either.

File: src/fsharp/Optimizer.fs (hunk 1 above)
Comment (TIHan) [reply src/fsharp/Optimizer.fs:2978]: Yea, trying to get something passing...

**Comment by TIHan** [comment]
There was a lot more to this than originally thought. Partial application isn't the same as wrapping the application in a lambda because the arguments are on the partial app are eager. Therefore, we have to check for effects.

**Comment by TIHan** [comment]
This is almost done. Just need to update some tests.

**Comment by TIHan** [comment]
This is now ready.

**Comment by TIHan** [comment]
@cartermp To ensure this actually improves anything, we might need to find a good test case. I don't want us to potentially regress.

File: src/fsharp/Optimizer.fs [hunk 3]
```diff
@@ -2901,6 +2901,55 @@ and OptimizeApplication cenv env (f0, f0ty, tyargs, args, m) =
...
+                    let action = test a b c d e f g
+                    printfn "world"
+                    xs |> Array.iter action
+            *)
+            // In the example, we should not make a lambda around "test" in "test2" because "test" is fully applied but returns
+            //     a partial application.
+
+            // The currently passed arguments must not have any effects.
+            let argTys, retTy = stripFunTy g ety
+            if valInfo.NumCurriedArgs = args.Length + argTys.Length && not (ExprsHaveEffect g args) then
```
Comment (TIHan) [inline src/fsharp/Optimizer.fs:2932]: I think we can't exactly convert an application to a lambda when all args simply don't have effects.
The args must be values that are not mutable.

**Comment by TIHan** [comment]
This technically regresses:

```fsharp
open System
open System.Runtime.CompilerServices

[<MethodImpl(MethodImplOptions.NoInlining)>]
let printIt a b c d e f g (x: int) = 
    ()

let test1 a b c d e f g xs =
    xs |> Array.iter (printIt a b c d e f g)


let bench () =
    GC.Collect(2, GCCollectionMode.Forced)
    let xs = Array.zeroCreate 10000000
    let s = System.Diagnostics.Stopwatch.StartNew()
    test1 1 2 3 4 5 6 7 xs
    s.Stop()
    printfn "time: %A" s.Elapsed.TotalMilliseconds


[<EntryPoint>]
let main argv =
    for i = 0 to 100 do
        bench ()
    0
```
On my machine, `bench ()` is slower with this PR at this time. I think it's technically due to continual passing of all arguments on iteration rather than just the one(the element).

I don't think the approach of converting an app to a lambda works in this case. The arguments are considered eager and not lazy. This will be more involved.

We could follow a C# approach of creating a struct with backing state that would represent the lambda's closed values.

Closing this for now. But, at least it was an interesting experiment.

**Pull request closed without merging**